### PR TITLE
[MM-20432] Migrate components/channel_view module and associated tests to TypeScript

### PR DESCRIPTION
**Pull request by kcc343**
#### Summary
Migrate components/channel_view module and associated tests to TypeScript

I'm kind of new to TypeScript so I was not sure about the proper way to convert the code. Let me know if I need to change anything!

#### Ticket Link

Github Issue: Fixes https://github.com/mattermost/mattermost-server/issues/16142
JIRA ticket: Fixes https://mattermost.atlassian.net/browse/MM-20432

### CODE REVIEW

File: components/channel_view/channel_view.tsx [hunk 1]
```diff
@@ -2,7 +2,8 @@
 // See LICENSE.txt for license information.
 /* eslint-disable react/no-string-refs */
 
-import PropTypes from 'prop-types';
+// import PropTypes from 'prop-types';
```
Comment (ethervoid) [inline components/channel_view/channel_view.tsx:5]: Please remove these commented lines

File: components/channel_view/channel_view.tsx [hunk 2]
```diff
@@ -55,7 +63,7 @@ export default class ChannelView extends React.PureComponent {
         );
     }
 
-    static getDerivedStateFromProps(props, state) {
+    public static getDerivedStateFromProps(props: any, state: any) {
```
Comment (ethervoid) [inline components/channel_view/channel_view.tsx:66]: Props and state should be of `Props` and `State` type right?

File: components/channel_view/channel_view.tsx [hunk 3]
```diff
@@ -81,24 +89,25 @@ export default class ChannelView extends React.PureComponent {
 
         return null;
     }
-
-    constructor(props) {
+    public channelViewRef: any;
```
Comment (ethervoid) [inline components/channel_view/channel_view.tsx:92]: I'm seeing all the methods public and I think is not necessary right? also avoid using any and try to put the type here

File: components/channel_view/channel_view.tsx [hunk 4]
```diff
@@ -109,7 +118,7 @@ export default class ChannelView extends React.PureComponent {
         }
     }
 
-    componentDidUpdate(prevProps) {
+    public componentDidUpdate(prevProps: any) {
```
Comment (ethervoid) [inline components/channel_view/channel_view.tsx:121]: `Props` type

File: components/channel_view/channel_view.tsx [hunk 5]
```diff
@@ -81,24 +89,25 @@ export default class ChannelView extends React.PureComponent {
 
         return null;
     }
-
-    constructor(props) {
+    public channelViewRef: any;
+    public constructor(props: any) {
```
Comment (ethervoid) [inline components/channel_view/channel_view.tsx:93]: `Props` type

File: components/channel_view/index.ts [hunk 6]
```diff
@@ -21,19 +21,19 @@ import {goToLastViewedChannel} from 'actions/views/channel';
 import {setShowNextStepsView} from 'actions/views/next_steps';
 import {isOnboardingHidden, showNextSteps, showNextStepsTips} from 'components/next_steps_view/steps';
 
-import ChannelView from './channel_view.jsx';
+import ChannelView from './channel_view';
 
 // Temporary selector until getDirectTeammate is converted to be redux-friendly
 const getDeactivatedChannel = createSelector(
-    (state, channelId) => {
+    (state: any, channelId: any) => {
```
Comment (ethervoid) [inline components/channel_view/index.ts:28]: Same here in this file with the any types

**Review by ethervoid — CHANGES_REQUESTED** [review]
Thank you very much for your contribution. Some changes required

File: components/channel_view/index.ts [hunk 7]
```diff
@@ -20,20 +20,29 @@ import {TutorialSteps, Preferences} from 'utils/constants';
...
+    getProfiles: (page?: number, perPage?: number, options?: any) => ActionFunc;
+}
+
 // Temporary selector until getDirectTeammate is converted to be redux-friendly
 const getDeactivatedChannel = createSelector(
-    (state: any, channelId: any) => {
+    (state: GlobalState, channelId: string) => {
         return getDirectTeammate(state, channelId);
     },
     (teammate: any) => {
```
Comment (kcc343) [inline components/channel_view/index.ts:43]: I wasn't sure how to change this one to be not any. I think this is supposed to be UserProfile but it give me this error:

`No overload matches this call.
  The last overload gave the following error.
    Argument of type '(state: GlobalState, channelId: string) => {}' is not assignable to parameter of type 'ParametricSelector<unknown, unknown, UserProfile>[]'.
      Type '(state: GlobalState, channelId: string) => {}' is missing the following properties from type 'ParametricSelector<unknown, unknown, UserProfile>[]': pop, push, concat, join, and 27 more.ts(2769)`

If you are supposed to change this let me know!

**Comment by kcc343** [comment]
@ethervoid I did your requested changes but have a comment above for one of them for a problem I had (if it is a problem). Let me know if I need to make any more changes!

File: components/channel_view/index.ts (hunk 7 above)
Comment (ethervoid) [reply components/channel_view/index.ts:43]: Have you tried doing `(state: GlobalState, channelId: string) => getDirectTeammate(state, channelId)` ?

File: components/channel_view/channel_view.tsx [hunk 8]
```diff
@@ -16,33 +17,41 @@ import TutorialView from 'components/tutorial';
...
+    showNextStepsTips: boolean;
+    isOnboardingHidden: boolean;
+    showNextStepsEphemeral: boolean;
+    channelIsArchived: boolean;
+    viewArchivedChannels: boolean;
+    isCloud: boolean;
+    actions: {
+        goToLastViewedChannel: () => Promise<{data: boolean}>;
+        setShowNextStepsView: (show: boolean) => Action;
+        getProfiles: (page?: number, perPage?: number, options?: any) => ActionFunc;
```
Comment (ethervoid) [inline components/channel_view/channel_view.tsx:41]: We should change this any too

File: components/channel_view/channel_view.tsx [hunk 9]
```diff
@@ -16,33 +17,41 @@ import TutorialView from 'components/tutorial';
...
+    },
+};
+
+type State = {
+    channelId: string;
+    url: string;
+    focusedPostId?: string;
+    deferredPostView: any;
```
Comment (ethervoid) [inline components/channel_view/channel_view.tsx:49]: This any too

File: components/channel_view/index.ts (hunk 7 above)
Comment (ethervoid) [reply components/channel_view/index.ts:43]: Probably I`d change that any for `(teammate: Record<string, any>) => {` until the `utils.jsx` is moved to TS

File: components/channel_view/index.ts [hunk 10]
```diff
@@ -14,26 +14,38 @@ import {withRouter} from 'react-router-dom';
...
 import {isOnboardingHidden, showNextSteps, showNextStepsTips} from 'components/next_steps_view/steps';
+import {GlobalState} from 'types/store';
+
+import ChannelView from './channel_view';
 
-import ChannelView from './channel_view.jsx';
+type Actions = {
+    goToLastViewedChannel: () => Promise<{data: boolean}>;
+    setShowNextStepsView: (show: boolean) => Action;
+    getProfiles: (page?: number, perPage?: number, options?: any) => ActionFunc;
```
Comment (ethervoid) [inline components/channel_view/index.ts:35]: This any should be tackled too

File: components/channel_view/channel_view.tsx [hunk 11]
```diff
@@ -55,7 +64,7 @@ export default class ChannelView extends React.PureComponent {
         );
     }
 
-    static getDerivedStateFromProps(props, state) {
+    public static getDerivedStateFromProps(props: Props, state: State) {
```
Comment (ethervoid) [inline components/channel_view/channel_view.tsx:67]: Is this public needed?

**Review by ethervoid — CHANGES_REQUESTED** [review]
Some more changes

File: components/channel_view/channel_view.tsx (hunk 9 above)
Comment (kcc343) [reply components/channel_view/channel_view.tsx:49]: @ethervoid Hopefully this isn't a bother but I'm not sure how to tackle this one? I've tried changing this to `typeof DeferredRenderWrapper & hoistNonReactStatics.NonReactStatics<any, {}>` but then it creates this error on line 104:

`Type 'typeof DeferredRenderWrapper & NonReactStatics<any, {}>' is not assignable to type '((WrappedComponent: any, PreRenderComponent?: any) => typeof DeferredRenderWrapper & NonReactStatics<any, {}>) & NonReactStatics<any, {}>'.
  Type 'typeof DeferredRenderWrapper & NonReactStatics<any, {}>' is not assignable to type '(WrappedComponent: any, PreRenderComponent?: any) => typeof DeferredRenderWrapper & NonReactStatics<any, {}>'.
    Type 'typeof DeferredRenderWrapper & NonReactStatics<any, {}>' provides no match for the signature '(WrappedComponent: any, PreRenderComponent?: any): typeof DeferredRenderWrapper & NonReactStatics<any, {}>'.`

and this error on line 248:

`'DeferredPostView' cannot be used as a JSX component.
  Its return type 'typeof DeferredRenderWrapper & NonReactStatics<any, {}>' is not a valid JSX element.
    Type 'typeof DeferredRenderWrapper & NonReactStatics<any, {}>' is missing the following properties from type 'Element': type, props, key``

I was also trying to search if it was a custom type in the mattermost-redux library but there is not one that seems to fit this.

Comment (ethervoid) [reply components/channel_view/channel_view.tsx:49]: Ok, we can leave as any for now :)

**Review by ethervoid — APPROVED** [review]
LGTM! Thank you for the contribution :)

**Comment by jasonblais** [comment]
@catalintomai friendly reminder to help with review :)

**Review by catalintomai — APPROVED** [review]

**Review by lindy65 — APPROVED** [review]
Thanks @kcc343 - looks good to me 👍 

@ethervoid - please help add areas to smoke-test to the [jira ticket](https://mattermost.atlassian.net/browse/MM-20432)? Thanks!

**Comment by hanzei** [comment]
@ethervoid I didn't close https://mattermost.atlassian.net/browse/MM-20432 as QA test steps are missing. Could you please take care of that?

**Comment by ethervoid** [comment]
@hanzei done!

**Comment by hanzei** [comment]
Thanks @ethervoid :+1:

**Pull request merged**